### PR TITLE
Feature/iia 913 dynamically update field order updated logic

### PR DESCRIPTION
**Pull request by swaroopsalvi**

### CODE REVIEW

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/model/PatternField.java [hunk 1]
```diff
@@ -19,14 +19,18 @@
...
  * @param dataType tinkar data type
  * @param purpose purpose concept
  * @param meaning meaning concept
  * @param comments user comments
  */
-public record PatternField(int fieldOrder, String displayName, EntityFacade dataType,
+public record PatternField(String displayName, EntityFacade dataType,
                           EntityFacade purpose, EntityFacade meaning, String comments) {
 
+//    public PatternField withFieldOrder(int fieldOrder){
```
Comment (dholubek) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/model/PatternField.java:32]: can we delete this commented out code?

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PropertiesController.java [hunk 2]
```diff
@@ -139,7 +139,11 @@ private void initialize() {
             if (evt.getEventType() == SHOW_ADD_DEFINITION) {
                 currentEditPane = patternDefinitionPane; // must be available.
             } else if (evt.getEventType() == SHOW_EDIT_FIELDS) {
+                patternFieldsController.updateViewModel(patternFieldsViewModel -> {
+                    patternFieldsViewModel.setPropertyValue(TOTAL_EXISTING_FIELDS, evt.getTotalFields());
+                });
                 currentEditPane = patternFieldsPane;
+             //   setupFieldsPane(evt.getFields());
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PropertiesController.java:146]: Can we remove the statement and remove the function?

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java [hunk 3]
```diff
@@ -152,23 +155,53 @@ private void initialize() {
...
+        loadDataTypeComboBox();
+        loadFieldOrderOptions(totalExistingfields.get());
+
+    }
+
+    private void loadFieldOrderOptions(int totalFields){
+
+        // get the available dropdown options initially list will be empty.
+        ObservableList<Integer> fieldOrderOptions = patternFieldsViewModel.getObservableList(FIELD_ORDER_OPTIONS);
+        int optionsSize = fieldOrderOptions.size();
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java:188]: Can we make it even simpler?
```java

// assuming the bound items are already done. 
ObservableList<Integer> fieldOrderOptions = patternFieldsViewModel.getObservableList(FIELD_ORDER_OPTIONS);

// Clear list
fieldOrderOptions.clear();

// Create a stream of integers from 1 to (total field + 1)
IntStream.rangeClosed(1, totalFields+1) 
                .boxed() // Convert int to Integer
                .forEach(fieldOrderOptions::add);
                
// Select the last item as a default to the user                
fieldOrderComboBox.getSelectionModel().selectLast();

```

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java [hunk 4]
```diff
@@ -461,22 +494,22 @@ private void collectFormData() {
...
             return; // do not proceed.
         }
-
         //publish close env
         EvtBusFactory.getDefaultEvtBus().publish(patternFieldsViewModel.getPropertyValue(PATTERN_TOPIC),
                 new PatternPropertyPanelEvent(actionEvent.getSource(), CLOSE_PANEL));
 
-
+        //TODO This If block is not required but for some reason the bidirectional binding declared in initialize method is not working with FIELD_ORDER.
+        if(fieldOrderComboBox.getSelectionModel().getSelectedItem() != patternFieldsViewModel.getValue(FIELD_ORDER)){
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java:508]: Something is not wired up correctly. 
```java
// The code below should be in the initialize() method
ObjectProperty<Integer> fieldOrderProp = patternFieldsViewModel.getProperty(FIELD_ORDER);
fieldOrderProp.bind(fieldOrderComboBox.getSelectionModel().selectedItemProperty()); 
```
If you'll notice this is not bidirectional. The user is selecting an item that will sync.

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java [hunk 5]
```diff
@@ -152,23 +156,46 @@ private void initialize() {
...
+        // get the available dropdown options initially list will be empty.
+        ObservableList<Integer> fieldOrderOptions = patternFieldsViewModel.getObservableList(FIELD_ORDER_OPTIONS);
+        // Clear list
+        fieldOrderOptions.clear();
+        // Create a stream of integers from 1 to (total field + 1)
+        IntStream.rangeClosed(1, totalFields+1)
+                .boxed() // Convert int to Integer
+                .forEach(fieldOrderOptions::add);
+
+        fieldOrderComboBox.setItems(fieldOrderOptions); // Set the items in fieldOrder
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java:195]: This should already have been bound in the initialize() method.

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java [hunk 6]
```diff
@@ -485,7 +509,18 @@ public void onDone(ActionEvent actionEvent) {
...
+
+    public void updateViewModel(Consumer<PatternFieldsViewModel> viewModelConsumer) {
+        if (viewModelConsumer != null) {
+            viewModelConsumer.accept(getViewModel());
+        }
+    }
+
+    public PatternFieldsViewModel getViewModel(){
+        return patternFieldsViewModel;
+    }
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java:525]: Can we remove the method?
It isn't being used.

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternDetailsController.java [hunk 7]
```diff
@@ -262,23 +260,27 @@ public void onChanged(Change<? extends DescrName> change) {
...
+        ObservableList fieldsTilePaneList = fieldsTilePane.getChildren();
+        fieldsTilePaneList.addListener((ListChangeListener<? super VBox>) (listener) -> {
             while(listener.next()){
-                if(listener.wasAdded()) {
-                    PatternField patternField = listener.getAddedSubList().getFirst();
-                    //TODO the fieldNum code should be removed and replaced by patternField.fieldOrder();
-                    int fieldNum = fieldsTilePane.getChildren().size() + 1;
-                    fieldsTilePane.getChildren().add(createFieldEntry(patternField, fieldNum));
+                if(listener.wasAdded()){
+                    updateFieldValues(listener.getTo());
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternDetailsController.java:279]: Can you comment this line?
Is `getTo()` correct?

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java [hunk 8]
```diff
@@ -485,7 +509,18 @@ public void onDone(ActionEvent actionEvent) {
         );
 
         EvtBusFactory.getDefaultEvtBus().publish(patternFieldsViewModel.getPropertyValue(PATTERN_TOPIC),
-                new PatternFieldsPanelEvent(actionEvent.getSource(), PATTERN_FIELDS, patternField));
+                new PatternFieldsPanelEvent(actionEvent.getSource(), PATTERN_FIELDS, patternField, patternFieldsViewModel.getValue(FIELD_ORDER)));
+
         clearView(actionEvent);
     }
+
+    public void updateViewModel(Consumer<PatternFieldsViewModel> viewModelConsumer) {
```
[the remark this answers is not in the document]
Comment (carldea) [reply kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternFieldsController.java:517]: (Optional) Don't have to do this, it is just a suggestion.
Notes: Keep in mind that a JavaFX controller can have one to many view models. While most will have at least one main one used this method is okay. Another approach when creating a `JFXNode<Pane, MyController>` using `FXMLMvvmLoader.make(config)` the code can create a NamedVM (view model) to be injected (into controller) instead.
```java
 Config config = new Config();
 PatternFieldsViewModel pfvm = new PatternFieldsViewModel();
 config.addNamedVM(new NamedVM("patternFieldsViewModel", pfvm));
 JFXNode<Pane, PatternFieldsController> jfxNode = FXMLMvvmLoader.make(config);
 ... 
```
When using the MVVM(pattern) with JavaFX controllers it is good practice to communicate through ViewModels and remove the need of methods on a controller. In the early stages of development I created these convenient functions (methods) in an interface called `BasicController` to change view model properties.  However, in the spirit of MVVM, It is better Controller to be stupid and have less to no methods at all. If controller code really needs to know about something from the outside use the strategy above, or use an event bus.

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/model/PatternField.java [hunk 9]
```diff
@@ -19,14 +19,14 @@
 
 /**
  * object to capture data for a Pattern>Field
- * @param fieldOrder order in 1..n
+ * //@param fieldOrder order in 1..n
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/model/PatternField.java:22]: Can we remove the unused comment?

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternDetailsController.java [hunk 10]
```diff
@@ -262,23 +260,27 @@ public void onChanged(Change<? extends DescrName> change) {
...
         };
+
         EvtBusFactory.getDefaultEvtBus().subscribe(patternViewModel.getPropertyValue(PATTERN_TOPIC), PatternFieldsPanelEvent.class, patternFieldsPanelEventSubscriber);
 
-        //TODO This observable listener should refresh and reorder as per the inserted records field order.
-        ObservableList<PatternField> patternFieldList = patternViewModel.getObservableList(FIELDS_COLLECTION);
-        patternFieldList.addListener((ListChangeListener<? super PatternField>) (listener) -> {
+        //Listen to the changes int he fieldsTilePane and update the field numbers.
+        ObservableList fieldsTilePaneList = fieldsTilePane.getChildren();
+        fieldsTilePaneList.addListener((ListChangeListener<? super VBox>) (listener) -> {
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternDetailsController.java:276]: There is no need to use a specific type cast.

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternDetailsController.java [hunk 11]
```diff
@@ -262,23 +260,27 @@ public void onChanged(Change<? extends DescrName> change) {
...
+            fieldsTilePane.getChildren().add(fieldPosition, createFieldEntry(patternField, evt.getCurrentFieldOrder()));
         };
+
         EvtBusFactory.getDefaultEvtBus().subscribe(patternViewModel.getPropertyValue(PATTERN_TOPIC), PatternFieldsPanelEvent.class, patternFieldsPanelEventSubscriber);
 
-        //TODO This observable listener should refresh and reorder as per the inserted records field order.
-        ObservableList<PatternField> patternFieldList = patternViewModel.getObservableList(FIELDS_COLLECTION);
-        patternFieldList.addListener((ListChangeListener<? super PatternField>) (listener) -> {
+        //Listen to the changes int he fieldsTilePane and update the field numbers.
+        ObservableList fieldsTilePaneList = fieldsTilePane.getChildren();
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PatternDetailsController.java:275]: Can we define the ObservableList as follows?
```java
ObservableList<Node> fieldsTilePaneList = fieldsTilePane.getChildren();
```

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PropertiesController.java [hunk 12]
```diff
@@ -155,17 +142,18 @@ private void initialize() {
         patternDefinitionController = patternDefinitionControllerJFXNode.controller();
         patternDefinitionPane = patternDefinitionControllerJFXNode.node();
 
-
         // +-----------------------------------
         // ! Edit field(s) within a Pattern
         // +-----------------------------------
         Config fieldsConfig = new Config(PATTERN_FIELDS_FXML_URL)
+                .addNamedViewModel(new NamedVm("patternPropertiesViewModel", patternPropertiesViewModel))
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PropertiesController.java:149]: Can you remove this statement? It is not used.

File: kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PropertiesController.java [hunk 13]
```diff
@@ -117,6 +103,7 @@ public class PropertiesController {
     private Pane currentEditPane;
 
     private PatternDefinitionController patternDefinitionController;
+    // private DescriptionNameController descriptionNameController;
```
Comment (carldea) [inline kview/src/main/java/dev/ikm/komet/kview/mvvm/view/pattern/PropertiesController.java:106]: Can we remove commented out code?

**Review by carldea — APPROVED** [review]
Very nice job. 
There are still 2 minor comments to resolve, but overall code is following JavaFX conventions.

**Review by dholubek — APPROVED** [review]
looks goo to me

**Pull request merged**